retrieval/discovery/kubernetes/discovery.go

```diff
@@ -774,3 +802,195 @@ func nodeHostIP(node *Node) (net.IP, error) {
 	}
 	return nil, fmt.Errorf("host IP unknown; known addresses: %v", addresses)
 }
+
+////////////////////////////////////
+// Here there be dragons.         //
+// Pod discovery code lies below. //
+////////////////////////////////////
+
+func (kd *Discovery) podSources(pods map[string]*Pod) []string {
+	var podNames []string
+	for _, pod := range pods {
+		podNames = append(podNames, podSource(pod))
+	}
+	return podNames
+}
+
+func (kd *Discovery) updatePod(pod *Pod, eventType EventType) *config.TargetGroup {
+	kd.podsMu.Lock()
+	defer kd.podsMu.Unlock()
+
+	switch eventType {
+	case deleted:
+		return kd.deletePod(pod)
+	case added, modified:
+		return kd.addPod(pod)
+	}
+	return nil
+}
+
+func (kd *Discovery) getPods() (map[string]*Pod, string, error) {
+	res, err := kd.queryAPIServerPath(podsURL)
+	if err != nil {
+		return nil, "", fmt.Errorf("Unable to list Kubernetes pods: %s", err)
```

Comment:
Please start error strings (except in log lines) with lower-case letters: https://github.com/golang/go/wiki/CodeReviewComments#error-strings

Here and below. With the ones that have two "sentences", maybe then use a semicolon (and then lower case) instead of a period, since it would look weird otherwise if the error itself doesn't start capitalized.